osbuild: add Duration field to Status and parse it

lzap opened the pull request:
osbuild: remove JSON control character from tests

The control character is being trimmed by the parser, linters do not
like this and I was confused when working with the test data too. This
patch removes it, but there is one test case that actually sends the
character into the parser to ensure it works properly.

---

osbuild: add Duration field to Status and parse it

Osbuild now emits duration of each stage/module in its status messages.
This patch adds Duration field to Status struct and parses it from
the JSON input. The field is of type time.Duration and is set to 0
if the duration is not present in the JSON (for example in older
osbuild versions).

---

osbuild: remove broken message fields

Osbuild does not always provide message fields in progress updates, this
was parsed as "Stage " and "Pipeline " which does not make sense.

---

I am not sure about the last commit, can drop if I am missing something. @mvo5 

https://issues.redhat.com/browse/HMS-9311

mvo5 commented on pkg/osbuild/monitor_test.go:
```diff
@@ -14,9 +14,9 @@ import (
 	"github.com/osbuild/images/pkg/osbuild"
 )
 
-const osbuildMonitorLines_curl = `{"message": "source/org.osbuild.curl (org.osbuild.curl): Downloaded https://rpmrepo.osbuild.org/v2/mirror/public/f39/f39-x86_64-fedora-20231109/Packages/k/kpartx-0.9.5-2.fc39.x86_64.rpm\n", "context": {"origin": "org.osbuild", "pipeline": {"name": "source org.osbuild.curl", "id": "598849389c35f93efe2412446f5ca6919434417b9bcea040ea5f9203de81db2c", "stage": {}}, " ...
```
Removing the control character makes the test less realistic, this is especially problematic when we do multiline test. The other issue is that its no longer trivial to just copy and paste output from osbuild to generate test-cases. I would not do it and instead add a comment with context (like "\x1e" is included because this follows the json-seq rfc or something).

The fact that our jsonseq parser is navie and just strips is the control chars instead of properly checking is actually an issue (not a big one) but if we fix it we would then be in a weird situation because the tests would no longer work as they are not realistic anymore.

lzap replied:
The problem is, this character can only be escaped in regular quotes, it is not visible which is a good argument against it.

mvo5 commented on pkg/osbuild/monitor_test.go:
```diff
@@ -72,13 +72,14 @@ func TestScannerSimple(t *testing.T) {
 
-	r := bytes.NewBufferString(osbuildMontiorLines_subprogress)
+	// test with the control character
+	r := bytes.NewBufferString("\x1e" + osbuildMontiorLines_subprogress)
```
For multi-line strings this will give us a very strange half-json-seq and half-not

mvo5 commented on pkg/osbuild/monitor.go:
```diff
@@ -123,22 +123,28 @@ func (sr *StatusScanner) Status() (*Status, error) {
 		Duration:  time.Duration(status.Duration * float64(time.Second)),
 	}
 
+	if pipelineName != "" {
```
A test case for this change wold be nice. In what situations does osbuild not give us a pipeline and/or a stage name? What does the progress struct look like in this case? Would be great to make sure that we provide still soemthing vaguely nice looking to the user and/or fix osbuild (if it needs fixing).

mvo5 replied:
Aha, nice. Thanks for this extra context!

lzap commented:
Implemented a helper that prefix JSONSeq control characters automatically so it does not need to be in the codebase.

As was requested, added an additional test for empty input which creates some problems with Timestamp parsing which I solved.

achilleas-k commented:
> We could just load the test data from an external file and then even go:embed it for easier access

I like this and had the same thought while reviewing.

bcl reviewed (approved):
Looks ok to me, assuming @mvo5's issues are resolved.

lzap commented:
I believe so, thank you very much. CLI followup: https://github.com/osbuild/image-builder-cli/pull/303
